### ISSUE TITLE
Question concerning PartialFC implementation

### ISSUE THREAD

**Issue by zwacke** [deepinsight/insightface#1316]
Hi,
I have a quick question concerning lines 127-131 in /recognition/partial_fc/pytorch/partial_fc.py

                max_fc = torch.max(logits, dim=1, keepdim=True)[0]
                dist.all_reduce(max_fc, dist.ReduceOp.MAX)
                # Calculate exp(logits) and all-reduce
                logits_exp = torch.exp(logits - max_fc)

I did not find any hint in the paper. If I see it correctly, for each sample you subtract the largest logit score across devices, i.e. all logit scores are shifted so that the values are <= 0. Afaik, the cross-entropy loss is invariant to constant shifts, so I assume the motivation must have something to do with stability of numerical values when calling the exponential function, or am I missing out on something?

**Comment by mnikitin** [comment]
@zwacke 
Yes, it's about numerical stability. Check [exp-normalization trick](https://timvieira.github.io/blog/post/2014/02/11/exp-normalize-trick/)
